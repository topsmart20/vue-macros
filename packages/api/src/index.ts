export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
