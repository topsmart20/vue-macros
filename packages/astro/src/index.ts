import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
