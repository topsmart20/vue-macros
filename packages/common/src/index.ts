export * from './vue'
export * from './ast'
export * from './constants'
export * from './unplugin'
export { default as MagicString } from 'magic-string'
