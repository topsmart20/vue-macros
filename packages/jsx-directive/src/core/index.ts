import {
  type JSXAttribute,
  type JSXElement,
  type Node,
  type Program,
} from '@babel/types'
import {
  MagicString,
  REGEX_SETUP_SFC,
  babelParse,
  generateTransform,
  getLang,
  parseSFC,
  walkAST,
} from '@vue-macros/common'
import { transformVIf } from './v-if'
import { transformVFor } from './v-for'
import { transformVMemo } from './v-memo'

export type JsxDirectiveNode = {
  node: JSXElement
  attribute: JSXAttribute & { matched?: RegExpMatchArray | null }
  parent?: Node | null
}

export function transformJsxDirective(code: string, id: string) {
  const lang = getLang(id)
  let asts: {
    ast: Program
    offset: number
  }[] = []
  if (lang === 'vue' || REGEX_SETUP_SFC.test(id)) {
    const { scriptSetup, getSetupAst, script, getScriptAst } = parseSFC(
      code,
      id
    )
    if (script) {
      asts.push({ ast: getScriptAst()!, offset: script.loc.start.offset })
    }
    if (scriptSetup) {
      asts.push({ ast: getSetupAst()!, offset: scriptSetup.loc.start.offset })
    }
  } else if (['jsx', 'tsx'].includes(lang)) {
    asts = [{ ast: babelParse(code, lang), offset: 0 }]
  } else {
    return
  }

  const s = new MagicString(code)
  for (const { ast, offset } of asts) {
    if (!/\sv-(if|for|memo|once)/.test(s.sliceNode(ast, { offset }))) continue

    const vIfMap = new Map<Node, JsxDirectiveNode[]>()
    const vForNodes: JsxDirectiveNode[] = []
    const vMemoNodes: (JsxDirectiveNode & {
      vForAttribute?: JSXAttribute
    })[] = []
    walkAST<Node>(ast, {
      enter(node, parent) {
        if (node.type !== 'JSXElement') return

        let vIfAttribute
        let vForAttribute
        let vMemoAttribute
        for (const attribute of node.openingElement.attributes) {
          if (attribute.type !== 'JSXAttribute') continue
          if (
            ['v-if', 'v-else-if', 'v-else'].includes(`${attribute.name.name}`)
          )
            vIfAttribute = attribute
          if (attribute.name.name === 'v-for') vForAttribute = attribute
          if (['v-memo', 'v-once'].includes(`${attribute.name.name}`))
            vMemoAttribute = attribute
        }

        if (vIfAttribute) {
          if (!vIfMap.has(parent!)) vIfMap.set(parent!, [])
          vIfMap.get(parent!)?.push({
            node,
            attribute: vIfAttribute,
            parent,
          })
        }
        if (vForAttribute) {
          vForNodes.push({
            node,
            attribute: vForAttribute,
            parent: vIfAttribute ? undefined : parent,
          })
        }
        if (vMemoAttribute) {
          vMemoNodes.push({
            node,
            attribute: vMemoAttribute,
            parent: vForAttribute || vIfAttribute ? undefined : parent,
            vForAttribute,
          })
        }
      },
    })

    vIfMap.forEach((nodes) => transformVIf(nodes, s, offset))
    transformVFor(vForNodes, s, offset)
    transformVMemo(vMemoNodes, s, offset)
  }

  return generateTransform(s, id)
}
