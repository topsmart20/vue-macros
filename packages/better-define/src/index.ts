import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { transformBetterDefine } from './core'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()

  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = 'unplugin-vue-better-define'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (err: unknown) {
          this.warn(`${name} ${err}`)
          console.warn(err)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  }
)
