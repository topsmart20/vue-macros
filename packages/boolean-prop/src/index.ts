import { type Plugin } from 'vite'
import { type VuePluginApi, getVuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Options, transformBooleanProp } from './core/index'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options)
      )
    },
  }
}

export default {
  rollup,
  vite: rollup,
}
