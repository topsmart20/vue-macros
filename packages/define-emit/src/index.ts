import { createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  type MarkRequired,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { transformDefineEmit } from './core'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOption(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC, REGEX_VUE_SUB],
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options)
    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineEmit(code, id)
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
