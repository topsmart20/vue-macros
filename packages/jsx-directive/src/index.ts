import { createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  type MarkRequired,
  REGEX_NODE_MODULES,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { transformJsxDirective } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'version'>

function resolveOption(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC],
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version,
  }
}

const name = 'unplugin-jsx-vue-directive'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}) => {
    const options = resolveOption(userOptions)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformJsxDirective(code, id)
      },
    }
  }
)
