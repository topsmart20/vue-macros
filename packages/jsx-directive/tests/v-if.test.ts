import { describe } from 'vitest'
import { testFixtures } from 'packages/test-utils/src'
import { transformJsxDirective } from '../src/api'

describe('jsx-vue-directive', () => {
  describe('v-if', async () => {
    await testFixtures(
      import.meta.glob('./fixtures/v-if/*.{vue,jsx,tsx}', {
        eager: true,
        as: 'raw',
      }),
      (args, id, code) => {
        return transformJsxDirective(code, id)?.code
      }
    )
  })
})
