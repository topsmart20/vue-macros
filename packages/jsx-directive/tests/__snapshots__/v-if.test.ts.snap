// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-else.tsx 1`] = `
"const { foo = 2 } = defineProps<{
  foo: number
}>()

export default () => (
  <>
    { foo === 0 ? <div>0</div> :
    foo === 1 ? <div>1</div> :
    <div v-else={foo === 2}>2</div> }
  </>
)
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-else.vue 1`] = `
"<script setup lang=\\"tsx\\">
const { foo = 2 } = defineProps<{
  foo: number
}>()

defineRender(() => (
  <>
    { foo === 0 ? <div>0</div> :
    foo === 1 ? <div>1</div> :
    <div v-else={foo === 2}>2</div> }
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-else-if.tsx 1`] = `
"const { foo = 0 } = defineProps<{
  foo: number
}>()

export default () => (
  <>
    { foo === 0 ? <div>0</div> :
    foo === 1 ? <div>1</div> :
    foo === 2 ? <div>2</div> : '' }
  </>
)
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-else-if.vue 1`] = `
"<script setup lang=\\"tsx\\">
const { foo = 0 } = defineProps<{
  foo: number
}>()

defineRender(() => (
  <>
    { foo === 0 ? <div>0</div> :
    foo === 1 ? <div>1</div> : '' }
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-if.tsx 1`] = `
"const { foo } = defineProps<{
  foo: number
}>()

export default () => {
  return (
    <>
      { foo === 0 ? <div>0</div> : '' }
      { foo === 1 ? <div>1</div> : '' }
    </>
  )
}
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-if.vue 1`] = `
"<script setup lang=\\"tsx\\">
const { foo } = defineProps<{
  foo: number
}>()

defineRender(() => (
  <>
    { foo === 0 ? <div>0</div> : '' }
    { foo === 1 ? <div>1</div> : '' }
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-if-nested.tsx 1`] = `
"const { foo = 0 } = defineProps<{
  foo: number
}>()

export default () => (
  <>
    { foo === 0 ? <div>
      { foo === 0 ? <div>0-0</div> :
      foo === 1 ? <div>0-1</div> :
      <div v-else>0-2</div> }
    </div> : '' }
  </>
)
"
`;

exports[`jsx-vue-directive > v-if > ./fixtures/v-if/v-if-nested.vue 1`] = `
"<script setup lang=\\"tsx\\">
const { foo = 0 } = defineProps<{
  foo: number
}>()

defineRender(() => (
  <>
    { foo === 0 ? <div>
      { foo === 0 ? <div>0-0</div> :
      foo === 1 ? <div>0-1</div> :
      <div v-else>0-2</div> }
    </div> : '' }
  </>
))
</script>
"
`;
